Tests/SwiftSyntaxMacroExpansionTest/LexicalContextTests.swift

```diff
@@ -0,0 +1,298 @@
+      let name = lexicalContext.functionName(in: context)
+    else {
+      return #""<unknown>""#
+    }
+
+    return ExprSyntax("\(literal: name)")
+  }
+}
+
+public struct AllLexicalContextsMacro: DeclarationMacro {
+  public static func expansion(
+    of node: some FreestandingMacroExpansionSyntax,
+    in context: some MacroExpansionContext
+  ) throws -> [DeclSyntax] {
+    context.lexicalContext.compactMap { $0.as(DeclSyntax.self)?.trimmed }
+  }
+}
+
+final class LexicalContextTests: XCTestCase {
+  private let indentationWidth: Trivia = .spaces(2)
+
+  func testPoundFunction() {
+    assertMacroExpansion(
+      """
+      func f(a: Int, _: Double, c: Int) {
+        print(#function)
+      }
+
+      struct X {
+        var computed: String {
+          get {
+            #function
+          }
+        }
+
+        init(from: String) {
+          #function
+        }
+
+        subscript(a: Int) -> String {
+          #function
+        }
+
+        subscript(a a: Int) -> String {
+          #function
+        }
+      }
+
+      extension A {
+        static var staticProp: String = #function
+      }
+      """,
+      expandedSource: """
+        func f(a: Int, _: Double, c: Int) {
+          print(  "f(a:_:c:)")
+        }
+
+        struct X {
+          var computed: String {
+            get {
+              "computed"
+            }
+          }
+
+          init(from: String) {
+            "init(from:)"
+          }
+
+          subscript(a: Int) -> String {
+            "subscript(_:)"
+          }
+
+          subscript(a a: Int) -> String {
+            "subscript(a:)"
+          }
+        }
+
+        extension A {
+          static var staticProp: String =   "staticProp"
+        }
+        """,
+      macros: ["function": FunctionMacro.self],
+      indentationWidth: indentationWidth
+    )
+  }
+
+  func testAllLexicalContexts() {
+    assertMacroExpansion(
+      """
+      extension A {
+        struct B {
+          func f(a: Int, b: Int) {
+            class C {
+              @A subscript(i: Int) -> String {
+                func g() {
+                  #allLexicalContexts
+                }
+              }
+            }
+          }
+        }
+      }
+      """,
+      expandedSource: """
+        extension A {
+          struct B {
+            func f(a: Int, b: Int) {
+              class C {
+                @A subscript(i: Int) -> String {
+                  func g() {
+                    func g()
+                    @A subscript(i: Int) -> String
+                    class C {
+                    }
+                    func f(a: Int, b: Int)
+                    struct B {
+                    }
+                    extension A {
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+        """,
+      macros: ["allLexicalContexts": AllLexicalContextsMacro.self]
+    )
+  }
```

Comment:
Would it make sense to also add a test case for a closure?

Reply:
Sure.